### ISSUE TITLE
Feature Request | Ability to add frame filtering to allow recording, monitoring and observation of audio

### ISSUE THREAD

**Issue by iDevelopper** [dimitris-c/AudioStreaming#1]
Hello,

The project promises to be awesome! Bravo!

This is not an issue as I have not tested yet in one of my projects, but a question.

The main question is:

How to achieve this (from streamingKit), as your framework is not yet documented:
```
audioPlayer.appendFrameFilter(withName: "CustomEQFilter") { (channelsPerFrame, bytesPerFrame, frameCount, frames) in
// Here I update my VU-Meter class for display.
}
```
Thanks,

**Comment by dimitris-c** [comment]
Hey

This is not yet implemented, it's shouldn't be hard to implement though.
I'll explore the options, as this can be done as in StreamingKit or by installing a tap on the mainMixer of AVAudioEngine, the later might be a better option.

**Comment by iDevelopper** [comment]
Hi @dimitris-c ,
I also use this feature for recording:
```swift
    func record() {
        let outputUrl = URL(fileURLWithPath: NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]).appendingPathComponent("output.m4a")
        var audioFile: AVAudioFile? = nil
        self.audioPlayer.addFrameFilter(withName: "RecordFilter", afterFilterWithName: "CustomEQFilter") { (channelsPerFrame, bytesPerFrame, frameCount, frames) in

            let sampleRate: Double = 44100

            if audioFile == nil {
                let settings = [
                    AVFormatIDKey: kAudioFormatMPEG4AAC,
                    AVSampleRateKey: sampleRate,
                    AVNumberOfChannelsKey: channelsPerFrame
                    ] as [String : Any]

                do {
                    audioFile = try AVAudioFile(forWriting: outputUrl, settings: settings, commonFormat: .pcmFormatInt16, interleaved: true)
                } catch {
                    let error = error as NSError
                    print("Error: \(error.localizedDescription)")
                }
            }

            let format = AVAudioFormat(commonFormat: .pcmFormatInt16, sampleRate: sampleRate, channels: channelsPerFrame, interleaved: true)
            let buffer = AVAudioPCMBuffer(pcmFormat: format!, frameCapacity: frameCount)

            buffer?.frameLength = frameCount
            memmove(buffer?.int16ChannelData?[0], frames, Int(frameCount * bytesPerFrame))

            do {
                if let buffer = buffer {
                    try audioFile?.write(from: buffer)
                }
            } catch let error as NSError {
                print("Error: \(error)")
            }
        }
    }
```

**Comment by iDevelopper** [comment]
Hi @dimitris-c ,

What about this enhancement? Do you think you will implement this feature soon?

Thanks,

**Comment by dimitris-c** [comment]
Yeah I have this one planned will do once I get a bit free time

**Comment by iDevelopper** [comment]
Ping

**Comment by dimitris-c** [comment]
@iDevelopper See #18

**Comment by iDevelopper** [comment]
@dimitris-c,

I see! If I understood correctly, I have to do like this:
```
    init() {
        player = AudioPlayer(configuration: .init(enableLogs: true))
        player.delegate = self

        configureAudioSession()
        registerSessionEvents()
        
        let filterEntry = FilterEntry(name: "NewEntry") { pcmBuffer, time in
            print(("buffer: \(pcmBuffer) - time: \(time)"))
        }
        player.frameFiltering.add(entry: filterEntry)
    }
```

I just need to find how to find the equivalent of the channelsPerFrame, bytesPerFrame, frameCount, frames of STKStreamingKit variables:
```
audioPlayer.appendFrameFilter(withName: "CustomEQFilter") { (channelsPerFrame, bytesPerFrame, frameCount, frames) in
// Here I update my VU-Meter class for display.
}```

Thanks!

**Comment by dimitris-c** [comment]
The values you're referring can be retrieved from the output format of the main mixer node like this
```
let format = player.mainMixerNode.outputFormat(forBus: 0)
format.basicStreamDescription.mChannelsPerFrame
format.basicStreamDescription.mBytesPerFrame
```

**Comment by dimitris-c** [comment]
@iDevelopper I see you're also using `.pcmFormatInt16` as the common format. 
Note that this has change the common format is now `.pcmFormatFloat32` so be sure to use the `floatChannelData` value of the `AVAudioPCMBuffer`, if you need access to this.

**Comment by iDevelopper** [comment]
If I replace `.pcmFormatInt16` by `.pcmFormatFloat32` in the record() function then there is a crash while:
```
            //memmove(buffer?.int16ChannelData?[0], frames, Int(frameCount * bytesPerFrame))
            memmove(buffer?.int32ChannelData?[0], frames, Int(frameCount * bytesPerFrame))
```
```swift
    func record() {
        let outputUrl = URL(fileURLWithPath: NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]).appendingPathComponent(recordFileName)
        var audioFile: AVAudioFile? = nil
        self.audioPlayer.addFrameFilter(withName: "RecordFilter", afterFilterWithName: "CustomEQFilter") { (channelsPerFrame, bytesPerFrame, frameCount, frames) in
            
            let sampleRate: Double = 44100

            if audioFile == nil {
                let settings = [
                    //AVFormatIDKey: kAudioFormatMPEG4AAC,
                    AVFormatIDKey: kAudioFormatLinearPCM,
                    AVSampleRateKey: sampleRate,
                    AVNumberOfChannelsKey: channelsPerFrame
                ] as [String : Any]
                
                do {
                    //audioFile = try AVAudioFile(forWriting: outputUrl, settings: settings, commonFormat: .pcmFormatInt16, interleaved: true)
                    audioFile = try AVAudioFile(forWriting: outputUrl, settings: settings, commonFormat: .pcmFormatFloat32, interleaved: true)
                } catch {
                    let error = error as NSError
                    print("Error: \(error.localizedDescription)")
                }
            }
            
            //let format = AVAudioFormat(commonFormat: .pcmFormatInt16, sampleRate: sampleRate, channels: channelsPerFrame, interleaved: true)
            let format = AVAudioFormat(commonFormat: .pcmFormatFloat32, sampleRate: sampleRate, channels: channelsPerFrame, interleaved: true)
            let buffer = AVAudioPCMBuffer(pcmFormat: format!, frameCapacity: frameCount)
            
            buffer?.frameLength = frameCount
            //memmove(buffer?.int16ChannelData?[0], frames, Int(frameCount * bytesPerFrame))
            memmove(buffer?.int32ChannelData?[0], frames, Int(frameCount * bytesPerFrame)) **// What should I do here?**

            do {
                if let buffer = buffer {
                    try audioFile?.write(from: buffer)
                }
            } catch let error as NSError {
                print("Error: \(error)")
            }
        }
    }
```

**Comment by iDevelopper** [comment]
> The values you're referring can be retrieved from the output format of the main mixer node like this

Ok,

Can this values can be retrieved in pcmBuffer.format too?

**Comment by iDevelopper** [comment]
If I replace:
```swift
            memmove(buffer?.int16ChannelData?[0], frames, Int(frameCount * bytesPerFrame))
```
by:
```swift
            memmove(buffer?.floatChannelData?[0], frames, Int(frameCount * bytesPerFrame))
```

No crash, but I try to play the recording with STKStreaminKit AudioPlay, it is inaudible. Only weird sounds.

**Comment by dimitris-c** [comment]
Don't use `int32ChannelData` but `floatChannelData`, I suppose you've checked out the feature branch which includes the FrameFilter work.

Also note that you don't need to re-create the buffer and move use `memmove`...
The FilterCallback will return a AVAudioPCMBuffer so you can pass that in the `write(from:)` method of `AVAudioFile`

And yes the `format` value should be the same as the `mainMixerNode.outputFormat(forBus: 0)`

**Comment by dimitris-c** [comment]
The branch name is `feature/frame-filters`, experiment with the changes from this one

**Comment by iDevelopper** [comment]
I was talking about my actual code with STKStreamingKit.

Also I have checked the feature/frame-filters branch. But I have just tried for now the add filter feature not the recording one.

`.pcmFormatFloat32` does not work with STkStreamingKit, perhaps because this library is out of date?

**Comment by iDevelopper** [comment]
I will try to duplicate my project and replace STKStreamingKit by your framework...

**Comment by dimitris-c** [comment]
Any code or changes I've mentioned are only related to this library, not StreamingKit...

**Comment by iDevelopper** [comment]
Yes I have understand. Help, how to get the frameCount value, please? And the void *frames equivalent? You can see I am bad in Audio!...

**Comment by dimitris-c** [comment]
I believe all the information can be retrieved from the buffer property
`AVAudioPCMBuffer` extends `AVAudioBuffer` in which you can find the `audioBufferList` from there you can extract the values you want
For the `void *frames` equivalent would be `buffer.audioBufferList.pointee.mBuffers.mData`

Mind if I ask why the need for `frameCount`?

**Comment by iDevelopper** [comment]
Thanks,

I think I wouldn't need it anymore (frameCount). In fact for vu-meter? I use the FreeStreamer (FSFrequencyDomainAnalyzer) (https://github.com/muhku/FreeStreamer) which I modify a bit to show a graph. You can download my app or see the screenshots of my app on the App Store to see (https://apps.apple.com/fr/app/shout-radios/id1474971083#?platform=iphone). For recordings, I will also allow the user to modify the recording to trim, in a future release. I am using  EZAudioPlot for that, but deprecated. It would be nice to have a version in your awesome library, but that will be the subject of a new feature request.

**Comment by iDevelopper** [comment]
In any case, thank you very much for this library and for your availability!

**Comment by dimitris-c** [comment]
This is resolved by #18
